haskell/README.md

```diff
@@ -63,26 +109,26 @@ with lenses we can do this
 lexer' = lexer & input %~(take 2) & position +~1 & readPosition +~1 & ch .~ 'b'
 ```
 
-This seems trivial here but when if we have a record with many fields it becomes very verbose and hard to read particularly with nested records.
+This seems trivial here but when if we have a record with many fields it
+becomes very verbose and hard to read particularly with nested records.
 
-## [Data.ByteString.Char8](https://hackage.haskell.org/package/bytestring-0.11.4.0/docs/Data-ByteString-Char8.html)
+### [Data.ByteString.Char8](https://hackage.haskell.org/package/bytestring-0.11.4.0/docs/Data-ByteString-Char8.html)
 
-Haskell `String` type is actually a list of `char`.
-This is not very performant, e.g., `length` and indexing `(!!)` operations are O(N). Also, indexing is unsafe (can fail) with crafting or importing safe operations. With `ByteString` we have O(1) operations and built in safe indexing that return `Maybe` types (Option in Rust).
-
-## [Data.Maybe](https://hackage.haskell.org/package/base-4.18.0.0/docs/Data-Maybe.html)
-
-Used to unwrap `Maybe` type (with default in our case).
-
-## [Data.Char](https://hackage.haskell.org/package/base-4.18.0.0/docs/Data-Char.html)
-
-Used  for built in `isSpace` , `isDigit`, `isLetter`
+Haskell `String` type is actually a linked list of `char`. This is not very
+performant, e.g., `length` and indexing `(!!)` operations are O(N). Also,
+indexing is unsafe (can fail) with crafting or importing safe operations. With
+`ByteString` we have O(1) operations and built in safe indexing that return
+`Maybe` types (Option in Rust).
 
 ## [Control.Monad.Trans.State](https://hackage.haskell.org/package/transformers-0.6.1.0/docs/Control-Monad-Trans-State.html)
 
-since we are working with immutable variables, we have to keep passing the state around. You can see this in the `OCaml` version where `lexer` is being parsed around recursively and returned in a tuple?.
+since we are working with immutable variables, we have to keep passing the
```

Comment:
I would say that this is because our functions are pure, and not that the variables are immutable. Since we can't have side effects, we need to pass along the variables.

The immutability makes it so that we keep having to create new data besides our old data, and make sure that we use the right variable name.

Reply:
Yes, this can be rephrased. Having the lexer as both input and output is basically keeping state. Maybe show an example like `String -> (a, String)` could be rewritten as a `State String a`.

Reply:
For me it looks like it makes sense, but for a beginner it might make it confusing since they would expect that State is a function itself. But it actually needs runState to run it.

I think that it might make more sense to be able to view it as

```rust
fn next_token(lexer: Lexer) -> (Token, Lexer) {
    // Logic
}
```

would be written as

```rust
impl Lexer {
    fn next_token(&mut self) -> Token {
        // Logic
    }
}
```

but this isn't really true either, because State doesn't really mutate anything

Reply:
True, but I still prefer your explanation. Even if it's not equivalent, it's the same function written as the best practice for the different languages, and it makes it clear what functions the parts of the type definition has.